lib/cache/sqlite-cache-store.js

```diff
@@ -0,0 +1,597 @@
+ */
+class SqliteCacheStore {
+  #maxEntries = Infinity
+
+  #maxEntrySize = Infinity
+
+  #entryCount = 0
+
+  /**
+   * @type {((err: Error) => void) | undefined}
+   */
+  #errorCallback
+
+  #db = null
+
+  #getValuesQuery
+
+  #insertValueQuery
+
+  #updateBodyQuery
+
+  #deleteByIdStatement
+
+  #deleteExpiredValuesStatement
+
+  #deleteByOriginStatement
+
+  /**
+   * Ids of rows currently being written too and so can't be read from
+   * @type {Set<number>}
+   */
+  #readLocks = new Set()
+
+  /**
+   * Denotes a row can't be written to because there's readables reading from it
+   * Row id:number of readables open for that value
+   * @type {Map<number, number>}
+   */
+  #writeLocks = new Map()
+
+  /**
+   * @param {import('../../types/cache-interceptor.d.ts').default.SqliteCacheStoreOpts | undefined} opts
+   */
+  constructor (opts) {
+    if (opts) {
+      if (typeof opts !== 'object') {
+        throw new TypeError('SqliteCacheStore options must be an object')
+      }
+
+      if (opts.maxEntries !== undefined) {
+        if (
+          typeof opts.maxEntries !== 'number' ||
+          !Number.isInteger(opts.maxEntries) ||
+          opts.maxEntries < 0
+        ) {
+          throw new TypeError('SqliteCacheStore options.maxEntries must be a non-negative integer')
+        }
+        this.#maxEntries = opts.maxEntries
+      }
+
+      if (opts.maxEntrySize !== undefined) {
+        if (
+          typeof opts.maxEntrySize !== 'number' ||
+          !Number.isInteger(opts.maxEntrySize) ||
+          opts.maxEntrySize < 0
+        ) {
+          throw new TypeError('SqliteCacheStore options.maxEntrySize must be a non-negative integer')
+        }
+        this.#maxEntrySize = opts.maxEntrySize
+      }
+
+      if (opts.errorCallback !== undefined) {
+        if (typeof opts.errorCallback !== 'function') {
+          throw new TypeError('SqliteCacheStore options.errorCallback must be a function')
+        }
+        this.#errorCallback = opts.errorCallback
+      }
+    }
+
+    this.#db = new DatabaseSync(opts?.location ?? ':memory:')
+
+    // Create the table and the indexes
+    this.#db.exec(`
+      CREATE TABLE IF NOT EXISTS cacheInterceptor (
+        -- Things we care about
+        id INTEGER PRIMARY KEY AUTOINCREMENT,
+        origin TEXT NOT NULL,
+        method TEXT NOT NULL,
+        body TEXT NULL,
+        deleteAt INTEGER NOT NULL,
+
+        -- Things the interceptor cares about
+        statusCode INTEGER NOT NULL,
+        statusMessage TEXT NOT NULL,
+        rawHeaders TEXT NULL,
+        rawTrailers TEXT NULL,
+        vary TEXT NULL,
+        cachedAt INTEGER NOT NULL,
+        staleAt INTEGER NOT NULL
+      );
+
+      CREATE INDEX IF NOT EXISTS idx_cacheInterceptor_origin ON cacheInterceptor(origin);
+      CREATE INDEX IF NOT EXISTS idx_cacheInterceptor_method ON cacheInterceptor(method);
+      CREATE INDEX IF NOT EXISTS idx_cacheInterceptor_deleteAt ON cacheInterceptor(deleteAt);
+    `)
```

Comment:
I think we need to add a tiny bit of migrations for this.

Reply:
Wdym?

Reply:
I think you need to query SQLite for that specific table, ask if the fields you need are there, and if not remove it and create it from scratch. It's a cache, it's ok ti wipe. Also check what Node.js is doing for handling localStorage.

Reply:
Agree with @mcollina, something similar to `localStorage` can be beneficial

Reply:
> Agree with @mcollina, something similar to `localStorage` can be beneficial

IMHO we should not do that. `SQLiteCacheStore` stores multi-dimensions data, it is just bad to use a KV-like `localStorage` w/ serialized cache.

Reply:
> IMHO, we should not do that. We are storing multi-dimensional data. It is just bad to use a KV-like `localStorage` with serialized cache keys. It is also difficult to implement cache expiration and purge with a KV-like `localStorage`.
> 
> IMHO, just create another cache provider with/ `localStorage` then.

I think they were referencing the implementation of `localStorage`, not about using it. The implementation in node uses sqlite, https://github.com/nodejs/node/blob/7ddd2c2282968cdb17b541bb8fe07f039c429897/src/node_webstorage.cc#L103